File: android/src/main/kotlin/com/example/mopro_flutter_package/MoproFlutterPackagePlugin.kt
package com.example.mopro_flutter_package

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import uniffi.mopro.generateCircomProof
import uniffi.mopro.CircomProof
import uniffi.mopro.ProofLib

/** MoproFlutterPackagePlugin */
class MoproFlutterPackagePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    this.flutterPluginBinding = flutterPluginBinding
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "mopro_flutter_package")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "getApplicationDocumentsDirectory" -> {
        val context = flutterPluginBinding.applicationContext
        val directory = context.getExternalFilesDir(null)
        result.success(directory?.absolutePath)
      }
      "generateProof" -> {
        val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
            "ARGUMENT_ERROR",
            "Missing zkeyPath",
            null
        )

        val inputs =
            call.argument<String>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

        val res = generateCircomProof(zkeyPath, inputs, ProofLib.ARKWORKS)
        val proof: CircomProof = res.proof

        val proofList = listOf(
            mapOf(
                "x" to proof.a.x,
                "y" to proof.a.y
            ), mapOf(
                "x" to proof.b.x,
                "y" to proof.b.y
            ), mapOf(
                "x" to proof.c.x,
                "y" to proof.c.y
            )
        )

        // Return the proof and inputs as a map supported by the StandardMethodCodec
        val resMap = mapOf(
            "proof" to proofList,
            "inputs" to res.inputs
        )

        result.success(
            resMap
        )
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/MoproFlutterPackagePlugin.swift
import Flutter
import moproFFI
import UIKit

public class MoproFlutterPackagePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "mopro_flutter_package", binaryMessenger: registrar.messenger())
    let instance = MoproFlutterPackagePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "generateProof":
      guard let args = call.arguments as? [String: Any],
        let zkeyPath = args["zkeyPath"] as? String,
        let inputs = args["inputs"] as? String
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        // Call the function from mopro.swift
        let proofResult = try generateCircomProof(
          zkeyPath: zkeyPath, circuitInputs: inputs, proofLib: ProofLib.arkworks)
        let proof = proofResult.proof
        let inputs = proofResult.inputs

        let proofList: [[String: Any]] = [
          ["x": proof.a.x, "y": proof.a.y],
          ["x": proof.b.x, "y": proof.b.y],
          ["x": proof.c.x, "y": proof.c.y],
        ]

        let resMap: [String: Any] = [
          "proof": proofList,
          "inputs": inputs,
        ]

        // Return the proof and inputs as a map supported by the StandardMethodCodec
        result(resMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
